### PR TITLE
[4.0] [com_csp] [plg_system_httpheaders] Implement a hashbased whitelist for com_csp / plg_http_httpheaders

### PR DESCRIPTION
**Pull request by zero-24**
### Summary of Changes

Implement a hashbased whitelist for com_csp / plg_http_httpheaders as requested by @SniperSister 

### Testing Instructions

- apply this patch
- System -> Manage -> Content Security Policy -> Options
- Enable the detect mode.
- navigate the frontend and backend
- System -> Manage -> Content Security Policy
- publish all rules
- System -> Manage -> Content Security Policy -> Options
- change the mode to "Automatic"
- check the header set by the app using the browser console network tab (hint check the first request you can see for the header)
- System -> Manage -> Content Security Policy -> Options
- enable the new script hashes and style hashes option
- check the header again.

### Expected result

New options for script and style hashes implemented

### Actual result

No options for script and style hashes implemented

### Documentation Changes Required

This needs to be documented.

### Known limitation

Please note that this here is indeed feature complete in order to work out of the box any inline style / script needs to be converted to a script running from a file. This is not the case with at least the current backend template so there are still some errors in the console but 90% of the features still work the other just needs to be moved to pass the required inline scripts via the joomla api. This is also true for any other template or extension.

### Open points

-[ ] en-GB review of the new language strings

@brianteeman can you please take a look into the new language strings whether they makes sense or can be improved? Thanks.

### Acknowledgements

Thanks @SniperSister for suggesting the implementation of hashbased whitelists in order to overcome the limitations of the nonce and the usage of `unsafe-inline`.

### CODE REVIEW

File: administrator/language/en-GB/en-GB.com_csp.ini [hunk 1]
```diff
@@ -12,12 +12,18 @@ COM_CSP_CONTENTSECURITYPOLICY_MODE_AUTO="Automatic"
...
+; Please only change the URL in the following language string
+COM_CSP_CONTENTSECURITYPOLICY_SCRIPT_HASHES_ENABLED="<a href='https://scotthelme.co.uk/content-security-policy-an-introduction/#hash' target='_blank' rel='noopener noreferrer'>Script hashes</a>"
+COM_CSP_CONTENTSECURITYPOLICY_SCRIPT_HASHES_ENABLED_DESC="Enable the optional hashbased whitelist for scripts und inline scripts using a cryptographic hash for all scripts using the Joomla API. Specifying hashes makes a modern browser ignore 'unsafe-inline' which should still be set for older browsers without hash support."
+; Please only change the URL in the following language string
+COM_CSP_CONTENTSECURITYPOLICY_STYLE_HASHES_ENABLED="<a href='https://scotthelme.co.uk/content-security-policy-an-introduction/#hash' target='_blank' rel='noopener noreferrer'>Style hashes</a>"
+COM_CSP_CONTENTSECURITYPOLICY_STYLE_HASHES_ENABLED_DESC="Enable the optional hashbased whitelist for stylesheets und inline styles using a cryptographic hash for all styles using the Joomla API. Specifying hashes makes a modern browser ignore 'unsafe-inline' which should still be set for older browsers without hash support."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_csp.ini]:
```suggestion
COM_CSP_CONTENTSECURITYPOLICY_STYLE_HASHES_ENABLED_DESC="Enable the optional hash based whitelist for stylesheets and inline styles using a cryptographic hash for all styles using the Joomla API. Specifying hashes makes a modern browser ignore 'unsafe-inline' which should still be set for older browsers without hash support."
```

File: administrator/language/en-GB/en-GB.com_csp.ini [hunk 2]
```diff
@@ -12,12 +12,18 @@ COM_CSP_CONTENTSECURITYPOLICY_MODE_AUTO="Automatic"
...
 ; Please do not translate 'Content-Security-Policy' & 'Content-Security-Policy-Report-Only' in the following language string
 COM_CSP_CONTENTSECURITYPOLICY_REPORT_ONLY_DESC="Use the header 'Content-Security-Policy-Report-Only' instead of 'Content-Security-Policy'."
 ; Please do not translate the following language string
 COM_CSP_CONTENTSECURITYPOLICY_REPORT_ONLY="Report-Only"
+; Please only change the URL in the following language string
+COM_CSP_CONTENTSECURITYPOLICY_SCRIPT_HASHES_ENABLED="<a href='https://scotthelme.co.uk/content-security-policy-an-introduction/#hash' target='_blank' rel='noopener noreferrer'>Script hashes</a>"
+COM_CSP_CONTENTSECURITYPOLICY_SCRIPT_HASHES_ENABLED_DESC="Enable the optional hashbased whitelist for scripts und inline scripts using a cryptographic hash for all scripts using the Joomla API. Specifying hashes makes a modern browser ignore 'unsafe-inline' which should still be set for older browsers without hash support."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_csp.ini]:
```suggestion
COM_CSP_CONTENTSECURITYPOLICY_SCRIPT_HASHES_ENABLED_DESC="Enable the optional hash based whitelist for scripts and inline scripts using a cryptographic hash for all scripts using the Joomla API. Specifying hashes makes a modern browser ignore 'unsafe-inline' which should still be set for older browsers without hash support."
```

**Comment by brianteeman** [comment]
> Specifying hashes makes a modern browser ignore 'unsafe-inline'

Is that really an accurate description

**Comment by brianteeman** [comment]
At least according to the linked source https://scotthelme.co.uk/content-security-policy-an-introduction/#hash then thats not how it is supposed to work

**Comment by SniperSister** [comment]
@zero-24 is correct, a browser with CSP2 support will ignore an unsafe-inline directive if hashes and/or nonces are present. That’s a built-in compatibility layer in the specification.

**Comment by brianteeman** [comment]
Surely it only "ignores" it if the hashes are checked and found to be correct

> the browser will hash any inline JS or CSS and see if the digest matches a value found in the header. If it does, the content is safe for use.
The contents of any <script> tags on the page would then be hashed and compared to the value found in the CSP header. If the values matched, the script would be allowed to execute.

**Comment by zero-24** [comment]
> Is that really an accurate description

Can you please suggest an alternative description?

**Comment by brianteeman** [comment]
It depends if it does what you wrote or it does what Scott says you should do

**Comment by zero-24** [comment]
hmm my text is based on the chrome error message:

>[Report Only] Refused to execute inline script because it violates the following Content Security Policy directive: "script-src-elem 'self' 'unsafe-inline' 'sha256-tbnnT/BTiNbMb7tp2DUzXvspk56uv/T7ZMNODFB3MMo='  'unsafe-inline'". **Note that 'unsafe-inline' is ignored if either a hash or nonce value is present in the source list.**

And in chrome and FF this also works just that way, that is what i can tell.

File: plugins/system/httpheaders/httpheaders.php [hunk 3]
```diff
@@ -157,9 +157,80 @@ public static function getSubscribedEvents(): array
...
+		$headers = $this->app->getHeaders();
+
+		foreach ($headers as $id => $headerConfiguration)
+		{
+			if (strtolower($headerConfiguration['name']) === 'content-security-policy'
+				|| strtolower($headerConfiguration['name']) === 'content-security-policy-report-only')
+			{
+				$newHeaderValue = strtolower($headerConfiguration['value']);
+
+				if (is_array($scriptHashes))
```
Comment (SniperSister) [inline plugins/system/httpheaders/httpheaders.php]: How can $scriptHashes NOT be an array?

File: plugins/system/httpheaders/httpheaders.php [hunk 4]
```diff
@@ -157,9 +157,80 @@ public static function getSubscribedEvents(): array
...
+
+		// Replace the hashes in the csp header when set.
+		$headers = $this->app->getHeaders();
+
+		foreach ($headers as $id => $headerConfiguration)
+		{
+			if (strtolower($headerConfiguration['name']) === 'content-security-policy'
+				|| strtolower($headerConfiguration['name']) === 'content-security-policy-report-only')
+			{
+				$newHeaderValue = strtolower($headerConfiguration['value']);
```
Comment (SniperSister) [inline plugins/system/httpheaders/httpheaders.php]: Wouldn't this break any existing nonce or hash using uppercase characters?

File: plugins/system/httpheaders/httpheaders.php [hunk 5]
```diff
@@ -157,9 +157,80 @@ public static function getSubscribedEvents(): array
...
+				if (is_array($scriptHashes))
+				{
+					$newHeaderValue = str_replace('{script-hashes}', ' ' . implode(' ', $scriptHashes), $newHeaderValue);
+				}
+				else
+				{
+					$newHeaderValue = str_replace('{script-hashes}', '', $newHeaderValue);
+				}
+
+				if (is_array($styleHashes))
```
Comment (SniperSister) [inline plugins/system/httpheaders/httpheaders.php]: See above, styleHashes is always an array

File: plugins/system/httpheaders/httpheaders.php [hunk 6]
```diff
@@ -316,9 +389,22 @@ private function setCspHeader(): void
 			{
 				if (in_array($cspValue->directive, $this->nonceDirectives) && $nonceEnabled)
 				{
+					// Append the nonce
 					$cspValue->value = str_replace('{nonce}', "'nonce-" . $this->cspNonce . "'", $cspValue->value);
 				}
 
+				// Append the script hashes placeholder
+				if ($scriptHashesEnabled && substr($cspValue->directive, 0, 10) === 'script-src')
```
Comment (SniperSister) [inline plugins/system/httpheaders/httpheaders.php]: substr can be replaced with strpos($cspValue->directive, 'script-src') === 0

File: plugins/system/httpheaders/httpheaders.php (hunk 4 above)
Comment (zero-24) [reply plugins/system/httpheaders/httpheaders.php]: You are correct. That line can be removed completely.

File: plugins/system/httpheaders/httpheaders.php (hunk 3 above)
Comment (zero-24) [reply plugins/system/httpheaders/httpheaders.php]: This should have been replaced by an empty check.

File: plugins/system/httpheaders/httpheaders.php (hunk 5 above)
Comment (zero-24) [reply plugins/system/httpheaders/httpheaders.php]: This also should have been replaced by an empty check.

File: plugins/system/httpheaders/httpheaders.php (hunk 6 above)
Comment (zero-24) [reply plugins/system/httpheaders/httpheaders.php]: ok same is true for style-src a few lines lower. Thanks

**Comment by zero-24** [comment]
Thanks for the feedback it is implemented now.

File: plugins/system/httpheaders/httpheaders.php [hunk 7]
```diff
@@ -411,9 +497,22 @@ private function compileAutomaticCspHeaderRules(): array
 			// Append the random $nonce for the script and style tags if enabled
 			if (in_array($cspHeaderkey, $this->nonceDirectives) && $nonceEnabled)
 			{
+				// Append nonce
 				$cspHeaderValue = "'nonce-" . $this->cspNonce . "'" . $cspHeaderValue;
 			}
 
+			// Append the script hashes placeholder
+			if ($scriptHashesEnabled && substr($cspHeaderkey, 0, 10) === 'script-src')
```
Comment (Quy) [inline plugins/system/httpheaders/httpheaders.php]: Change to `strpos`?

Comment (zero-24) [reply plugins/system/httpheaders/httpheaders.php]: Fixed thanks.

**Comment by wilsonge** [comment]
Thanks!

**Pull request merged**